sql-plugin/src/main/java/com/nvidia/spark/rapids/iceberg/spark/source/GpuMultiFileBatchReader.java

```diff
@@ -250,4 +173,278 @@ private Schema requiredSchema(GpuDeleteFilter deleteFilter) {
+        Map<Integer, ?> idToConsts = constsSchema._1();
+        Schema updatedReadSchema = constsSchema._2();
+        return GpuIcebergReader.addUpcastsIfNeeded(
+            GpuIcebergReader.addConstantColumns(batch, updatedReadSchema, idToConsts),
+            updatedReadSchema);
+      }
+    }
+  }
+
+  class ParquetCoalescingBatchReader extends MultiFileBatchReaderBase {
+
+    ParquetCoalescingBatchReader() {
+      super();
+    }
+
+    @Override
+    protected FilePartitionReaderBase createRapidsReader(PartitionedFile[] pFiles,
+        StructType partitionSchema) {
+      ArrayList<ParquetSingleDataBlockMeta> clippedBlocks = new ArrayList();
+      files.values().forEach(fst -> {
+        FilteredParquetFileInfo filteredInfo = filterParquetBlocks(fst);
+        List<ParquetSingleDataBlockMeta> fileSingleMetas =
+          JavaConverters.asJavaCollection(filteredInfo.parquetBlockMeta.blocks()).stream()
+            .map(b -> ParquetSingleDataBlockMeta.apply(
+                filteredInfo.parquetBlockMeta.filePath(),
+                ParquetDataBlock.apply(b),
+                InternalRow.empty(),
+                ParquetSchemaWrapper.apply(filteredInfo.parquetBlockMeta.schema()),
+                filteredInfo.parquetBlockMeta.readSchema(),
+                new IcebergParquetExtraInfo(
+                    filteredInfo.parquetBlockMeta.isCorrectedRebaseMode(),
+                    filteredInfo.parquetBlockMeta.isCorrectedInt96RebaseMode(),
+                    filteredInfo.parquetBlockMeta.hasInt96Timestamps(),
+                    filteredInfo.idToConstant(),
+                    filteredInfo.expectedSchema(),
+                    fst.spec())))
+            .collect(Collectors.toList());
+        clippedBlocks.addAll(fileSingleMetas);
+      });
+
+      return new MultiFileParquetPartitionReader(conf, pFiles,
+          JavaConverters.asScalaBuffer(clippedBlocks).toSeq(),
+          caseSensitive, parquetDebugDumpPrefix, maxBatchSizeRows, maxBatchSizeBytes,
+          metrics, partitionSchema, numThreads,
+          false, // ignoreMissingFiles
+          false, // ignoreCorruptFiles
+          false // useFieldId
+      ) {
+        @Override
+        public boolean checkIfNeedToSplitDataBlock(SingleDataBlockInfo currentBlockInfo,
+            SingleDataBlockInfo nextBlockInfo) {
+          // Check the read schema. Because it may differ among files in Iceberg.
+          if (!TrampolineUtil.sameType(currentBlockInfo.readSchema(),
+              nextBlockInfo.readSchema())) {
+            return true;
+          }
+          // Now for Iceberg, blocks with different partition schemas or partition values
+          // do not coalesce.
+          // Will try to figure out if it is possible to merge and add different
+          // partition values correctly in the future, to allow coalescing even
+          // partition values differ but with the same partition schema.
```

Comment:
Nit: Would be nice to link the spark-rapids issue here.

Reply:
updated